dot/parachain/collator-protocol/message.go

```diff
@@ -177,6 +178,227 @@ const (
+// validated.
+func (cpvs CollatorProtocolValidatorSide) enqueueCollation(
+	collations Collations,
+	relayParent common.Hash,
+	paraID parachaintypes.ParaID,
+	peerID peer.ID,
+	collatorID parachaintypes.CollatorID,
+	prospectiveCandidate *ProspectiveCandidate) error {
+
+	// TODO: return errors
+	pendingCollation := PendingCollation{
+		RelayParent:          relayParent,
+		ParaID:               paraID,
+		PeerID:               peerID,
+		ProspectiveCandidate: prospectiveCandidate,
+	}
+
+	switch collations.status {
+	// TODO: In rust code, a lot of thing that are being done in handle_advertisement
+	// are being repeated here.
+	// Currently enqueueCollation is being called from handle_advertisement only, so we might not need to
+	// repeat that here.
+	// If enqueueCollation gets used somewhere else, we would need to repeat those things here.
+
+	case Fetching, WaitingOnValidation:
+		logger.Debug("added collation to unfetched list")
+		collations.waitingQueue = append(collations.waitingQueue, UnfetchedCollation{
+			CollatorID:       collatorID,
+			PendingCollation: pendingCollation,
+		})
+	case Waiting:
+		// limit is not reached, it's allowed to second another collation
+		return cpvs.fetchCollation(pendingCollation)
+	case Seconded:
+		perRelayParent, ok := cpvs.perRelayParent[relayParent]
+		if !ok {
+			logger.Error("candidate relay parent went out of view for valid advertisement")
+			return ErrRelayParentUnknown
+		}
+		if perRelayParent.prospectiveParachainMode.isEnabled {
+			return cpvs.fetchCollation(pendingCollation)
+		} else {
+			logger.Debug("a collation has already been seconded")
+		}
+	}
+
+	return nil
+}
+
+func (cpvs *CollatorProtocolValidatorSide) fetchCollation(pendingCollation PendingCollation) error {
+	var candidateHash *parachaintypes.CandidateHash
+	if pendingCollation.ProspectiveCandidate != nil {
+		candidateHash = &pendingCollation.ProspectiveCandidate.CandidateHash
+	}
+	peerData, ok := cpvs.peerData[pendingCollation.PeerID]
+	if !ok {
+		return ErrUnknownPeer
+	}
+
+	if !peerData.HasAdvertised(pendingCollation.RelayParent, candidateHash) {
+		return ErrNotAdvertised
+	}
+
+	// TODO: Add it to collation_fetch_timeouts if we can't process this in timeout time.
+	// state
+	// .collation_fetch_timeouts
+	// .push(timeout(id.clone(), candidate_hash, relay_parent).boxed());
+	collation, err := cpvs.requestCollation(pendingCollation.RelayParent, pendingCollation.ParaID,
+		pendingCollation.PeerID)
+	if err != nil {
+		return fmt.Errorf("requesting collation: %w", err)
+	}
+
+	cpvs.fetchedCollations = append(cpvs.fetchedCollations, *collation)
+
+	return nil
+}
+
+func (cpvs *CollatorProtocolValidatorSide) handleAdvertisement(relayParent common.Hash, sender peer.ID,
+	prospectiveCandidate *ProspectiveCandidate) error {
+	// TODO:
+	// - tracks advertisements received and the source (peer id) of the advertisement
+	// - accept one advertisement per collator per source per relay-parent
+
```

Comment:
Looking at the polkadot code I see that there is a signature verification that happens during this process: https://github.com/paritytech/polkadot-sdk/blob/f4827dcd16fc5585a8d15c6a32bc27ecde00e7f9/polkadot/node/network/collator-protocol/src/validator_side/mod.rs#L810C3-L810C91 

Do we need to do something like this as well?

Reply:
The link you have shared belongs to logic of processing `Declare` message, not `AdvertiseCollation` message.

In processing `Declare` message we have verified the signature at this permalink
https://github.com/ChainSafe/gossamer/blob/496560dd314d6095501935f9e632d70b1004af5f/dot/parachain/collator-protocol/message.go#L483